File: EntityLib/TestEntGen/test_entgen.cpp
#include <EntGen.h>
#include <ciso646>
#include <filesystem>
#include <iostream>

using namespace Ent::Gen;

int main()
try
{
    Ent::EntityLib entlib("X:/", true);
    entlib.rawdataPath =
        std::filesystem::canonical(std::filesystem::current_path() / "../Test");

    auto entNode = entlib.loadEntityAsNode("instance.entity");
    auto ent = Entity(entNode.get());

    // Test Object
    auto name = ent.Name(); // inferred type : String
    static_assert(std::is_same_v<decltype(name), String>);
    auto components = ent.Components(); // inferred type : Components
    // Object Load
    auto ent2node = Entity::load(entlib, "instance.entity");
    auto ent2 = Entity(ent2node.get());
    // Object save
    ent2.save("instance_test_save.entity");

    // Test Set of Union
    auto ss = components.get<SubScene>(); // inferred type : SubScene
    for (Ent::String const& key : components.getKeys())
    {
        auto comp = components.get(key.c_str());
        std::cout << comp->getType() << std::endl;
    }
    for (Component const& comp : components)
    {
        std::cout << comp.getType() << std::endl;
    }
    auto tgd =
        ent.Components().get<TransformGD>(); // inferred type : std::optional<TransformGD>
    auto heightObj = components.get<HeightObj>();
    static_assert(
        std::is_same_v<std::remove_reference_t<decltype(*heightObj)>, HeightObj>);
    auto actorStateCreature = ent.ActorStates().add<ActorStateBeingImmersed>();
    auto unitTestCpt =
        components.add<UnitTestComponent>(); // inferred type : TurretComponentGD
    static_assert(std::is_same_v<decltype(unitTestCpt), UnitTestComponent>);
    heightObj = *components.HeightObj(); // Keep this method??

    // Test fixed size array
    auto pos =
        ent.Components().get<TransformGD>()->Position(); // inferred type : Vector3
    auto x = tgd->Position()[0]; // inferred type : Float
    // Copy point into Node
    double arr[] = {1., 2., 3.};
    tgd->Position() = std::vector<double>{1., 2., 3.};
    tgd->Position() = std::array<double, 3>{1., 2., 3.};
    tgd->Position() = std::tuple{1., 2., 3.};
    tgd->Position() = std::initializer_list<double>{1., 2., 3.};
    tgd->Position() = arr;
    tgd->Position()[1] = 2.;
    // Copy Node into point
    std::vector<double> v = tgd->Position();
    std::array<double, 3> a = tgd->Position();
    std::tuple<double, double, double> t = tgd->Position();

    // Test set of primitive
    auto side =
        unitTestCpt.EnumSet(); // inferred type : PrimitiveSet<ReviveSideTargetedItem_ReviveSide>
    static_assert(std::is_same_v<decltype(side), PrimitiveSet<ReviveSideEnum>>);
    unitTestCpt.EnumSet().add(ReviveSideEnum::neutral);
    unitTestCpt.EnumSet().add(ReviveSideEnum::cursed);
    // turret.ReviveSideTargeted.remove(ReviveSideEnum::neutral);
    ENTLIB_ASSERT(unitTestCpt.EnumSet().count(ReviveSideEnum::cursed));
    unitTestCpt.EnumSet().add(ReviveSideEnum::sacred); // inferred type : String
    for (auto&& enm : unitTestCpt.EnumSet())
    {
        std::cout << toString(enm) << std::endl;
        ENTLIB_ASSERT(unitTestCpt.EnumSet().count(enm));
    }
    // Test Enum prop
    auto mode =
        ent.Components().add<BeamTargetGD>().Mode(); // inferred type : BeamStaffMode
    mode = BeamStaffModeEnum::Hatching;
    BeamStaffModeEnum enm = mode;
    ENTLIB_ASSERT(enm == BeamStaffModeEnum::Hatching);

    // Test primitive types
    ENTLIB_ASSERT(ent.Name().get() == std::string("PlayerSpawner_"));
    ent.Name() = std::string("PlayerSpawner_2");
    ent.Name().set(std::string("PlayerSpawner_2"));
    ENTLIB_ASSERT(ent.Name().get() == std::string("PlayerSpawner_2"));
    ent.Name() = "PlayerSpawner_3";
    ENTLIB_ASSERT(ent.Name().isSet());
    ent.Name().unset();
    ENTLIB_ASSERT(not ent.Name().isSet());
    ENTLIB_ASSERT(heightObj->Subdivision().get() == 0);
    auto displaceNoise0 =
        heightObj->DisplaceNoiseList()[0]; // inferred type : DisplaceNoiseListItem
    ENTLIB_ASSERT(ent.Components().get<PathNodeGD>()->Radius().get() == 30.0);
    ENTLIB_ASSERT(ent.Components().get<PathNodeGD>()->Radius().toFloat() == 30.0f);
    [[maybe_unused]] float rad2 =
        ent.Components().get<PathNodeGD>()->Radius().toFloat();
    ent.Components().get<PathNodeGD>()->Radius() = 20.0;
    double rad = ent.Components().get<PathNodeGD>()->Radius();
    ENTLIB_ASSERT(rad == 20.0);

    // Test Tuple
    auto position = unitTestCpt.Position(); // inferred type : Position
    auto elt2_ = position.get<2>(); // inferred type : Float
    static_assert(std::is_same_v<decltype(elt2_), Float>);
    auto val0 = position.get<0>().get(); // inferred type : int
    static_assert(std::is_same_v<decltype(val0), int64_t>);
    ENTLIB_ASSERT(position.size() == 5);

    // Test Set of Object
    auto subscene =
        ent.Components().get<SubScene>()->Embedded(); // inferred type : ObjectSet<Entity>
    auto subent = subscene["EntityWithInstanceOf"]; // inferred type : Entity
    static_assert(std::is_same_v<decltype(subent), std::optional<Entity>>);
    auto new_ent = subscene.add("NewEntity"); // inferred type : Entity
    static_assert(std::is_same_v<decltype(new_ent), Entity>);
    for (auto&& entname : subscene.getKeys())
    {
        auto ent3 = subscene.get(entname.c_str());
        static_assert(std::is_same_v<decltype(ent3), std::optional<Entity>>);
    }
    for (auto ent3 : subscene)
    {
        static_assert(std::is_same_v<decltype(ent3), Entity>);
    }

    // Test Map
    auto pathNodeGD =
        ent.Components().get<PathNodeGD>(); // inferred type : PathNodeGD
    auto tags =
        pathNodeGD->Tags().Tags(); // inferred type : Map<char const*, PrimitiveSet<String>>
    static_assert(
        std::is_same_v<decltype(tags), Map<char const*, PrimitiveSet<char const*>>>);
    auto b = tags.get("b"); // inferred type : PrimitiveSet<String>
    tags.remove("c");
    auto c = tags.add("c"); // inferred type : PrimitiveSet<String>
    for (auto&& mapName : tags.getKeys())
    {
        auto set = tags.get(mapName.c_str());
        static_assert(
            std::is_same_v<decltype(set), std::optional<PrimitiveSet<char const*>>>);
    }
    for (auto&& [setname, set] : tags)
    {
        static_assert(std::is_same_v<decltype(set), PrimitiveSet<char const*>>);
    }

    // Test Map with enum key
    Map<RegenerationStateEnum, Array<sEnvStamp>> stamps =
        ent.Components().add<EnvStampGD>().Stamps();
    stamps.add(RegenerationStateEnum::Dead);
    stamps.add(RegenerationStateEnum::Dead);
    std::optional<Array<sEnvStamp>> deadArr =
        stamps.get(RegenerationStateEnum::Dead);
    ENTLIB_ASSERT(deadArr.has_value());
    std::optional<Array<sEnvStamp>> lushArr =
        stamps.get(RegenerationStateEnum::Lush);
    ENTLIB_ASSERT(not lushArr.has_value());
    for (auto&& stampname : stamps.getKeys())
    {
        auto stamp = stamps.get(stampname);
        std::cout << toString(stampname) << std::endl;
        static_assert(
            std::is_same_v<decltype(stamp), std::optional<Array<sEnvStamp>>>);
    }
    for (auto&& [stampname, stamp] : stamps)
    {
        std::cout << toString(stampname) << std::endl;
        static_assert(std::is_same_v<decltype(stamp), Array<sEnvStamp>>);
    }
    stamps.remove(RegenerationStateEnum::Dead);
    ENTLIB_ASSERT(not stamps.get(RegenerationStateEnum::Dead).has_value());
    return EXIT_SUCCESS;
}
catch (std::exception& ex)
{
    fprintf(stderr, ex.what());
    return EXIT_FAILURE;
}
catch (...)
{
    fprintf(stderr, "Exception thrown");
    return EXIT_FAILURE;
}

File: EntityLib/EntLibAPIGenerator/test_entgen.py
import os
import entgen_helpers
# from entgen.inline import *
from entgen._all import *
"""
from entgen.Object import *
from entgen.String import *
from entgen.SubScene import *
from entgen.TransformGD import *
from entgen.HeightObj import *
from entgen.ActorStateBeingImmersed import *
from entgen.UnitTestComponent import *
from entgen.BeamTargetGD import *
from entgen.PathNodeGD import *
from entgen.EnvStampGD import *
from entgen.AnimationControllerGD import *
"""

import EntityLibPy as Ent

Entity = Object

entlib = Ent.EntityLib("X:/")
entlib.rawdata_path = Ent.path(os.path.normpath(os.getcwd() + "/../Test"))

ent = Entity(entlib.load_node_entity(Ent.path("instance.entity")))

# Test Object
name = ent.Name  # inferred type : String
assert isinstance(name, String)
components = ent.Components  # inferred type : UnionSet
# load Object
ent2node = Entity.load(entlib, "instance.entity")
ent2 = Entity(ent2node)
# save Object
ent2.save("instance_test_save.entity")

# Test Set of Union
ss = components.get(SubScene)  # inferred type : SubScene
for key in components.keys():
    comp = components.get_by_name(key)
    assert isinstance(key, str)
    print(comp.schema.name)
for comp in components:
    assert isinstance(comp, EntityLibPy.Node)
    print(comp.schema.name)
tgd = ent.Components.get(TransformGD)  # inferred type : TransformGD
acgd = ent.Components.get(AnimationControllerGD)  # inferred type : TransformGD
assert acgd is None
heightObj = components.get(HeightObj)
assert isinstance(heightObj, HeightObj)
actorStateCreature = ent.ActorStates.add(ActorStateBeingImmersed)
unitTestCpt = components.add(UnitTestComponent)   # inferred type : UnitTestComponent
assert isinstance(unitTestCpt, UnitTestComponent)

# Test fixed size array
pos = ent.Components.get(TransformGD).Position  # inferred type : Array[Float]
x = tgd.Position[0]                             # inferred type : Float
# Copy point into Node double
tgd.Position = (1., 2., 3.)
tgd.Position = [1., 2., 3.]
# Copy Node into point
v = tgd.Position.value
assert v == [1., 2., 3.]
tgd.Position.value = [4., 5., 6.]
v = tgd.Position.value
assert v == [4., 5., 6.]

# Test set of primitive
side = unitTestCpt.EnumSet   # inferred type : PrimitiveSet[ReviveSideEnum]
assert isinstance(side, PrimitiveSet)
unitTestCpt.EnumSet.add(ReviveSideEnum.neutral)
unitTestCpt.EnumSet.add(ReviveSideEnum.cursed)
# unitTestCpt.EnumSet.remove("neutral")
assert unitTestCpt.EnumSet.count(ReviveSideEnum.cursed)
# unitTestCpt.EnumSet.erase("cursed")
unitTestCpt.EnumSet.add(ReviveSideEnum.sacred)  # inferred type : String
# del unitTestCpt.EnumSet["sacred"]
for enm in unitTestCpt.EnumSet:
    print(enm)

# Test Enum prop
# Do we want that a property of type BeamStaffModeEnum:
#     ent.Components.add(BeamTargetGD).Mode = BeamStaffModeEnum.Hatching
#     mode = ent.Components.add(BeamTargetGD).Mode  # return BeamStaffModeEnum
#     assert mode == BeamStaffModeEnum.Hatching
# Or do we want an object to manipulate the Node:
mode = ent.Components.add(BeamTargetGD).Mode  # inferred type: BeamStaffMode
assert isinstance(mode, BeamStaffMode)
mode.value = BeamStaffModeEnum.Hatching
assert isinstance(mode, BeamStaffMode)
enm = mode.value
assert enm is BeamStaffModeEnum.Hatching


# Test primitive types
assert ent.Name.value == "PlayerSpawner_"
assert ent.Name.get() == "PlayerSpawner_"
# ent.Name = "PlayerSpawner_2"  # Not possible
ent.Name.set("PlayerSpawner_3")
assert ent.Name.get() == "PlayerSpawner_3"
ent.Name.value = "PlayerSpawner_2"
assert ent.Name.get() == "PlayerSpawner_2"
ent.Name = "PlayerSpawner_4"
assert ent.Name.get() == "PlayerSpawner_4"
assert ent.Name.is_set() is True
ent.Name.unset()
assert ent.Name.is_set() is False
assert heightObj.Subdivision.value == 0
displaceNoise0 = heightObj.DisplaceNoiseList[0]  # inferred type : DisplaceNoiseListItem
assert ent.Components.get(PathNodeGD).Radius.value == 30.0

# Test Tuple
position = unitTestCpt.Position   # inferred type : Position
elt2 = position[2]           # inferred type : Any   :‑(     (waiting for python 3.9)
assert isinstance(elt2, Float)
elt2_ = position.get_2()                 # inferred type : Float
assert isinstance(elt2_, Float)
val0 = position.get_0().value            # inferred type : int
assert isinstance(val0, int)
assert len(position) == 5

# Test Set of Object
subscene = ent.Components.get(SubScene).Embedded   # inferred type : ObjectSet[Object]
subent = subscene["EntityWithInstanceOf"]          # inferred type : Object
assert isinstance(subent, Entity)
new_ent = subscene.add("NewEntity")                # inferred type : Object
assert isinstance(new_ent, Entity)
for entname in subscene.keys():
    ent3 = subscene.get(entname)
    assert isinstance(ent3, Entity)
for ent3 in subscene:
    assert isinstance(ent3, Entity)

# Test Map
pathNodeGD = ent.Components.get(PathNodeGD)   # inferred type : PathNodeGD
tags = pathNodeGD.Tags.Tags     # inferred type : Map[str, PrimitiveSet[String]]
assert isinstance(tags, Map)
b = tags.get("b")               # inferred type : PrimitiveSet[String]
tags.remove("c")
c = tags.add("c")               # inferred type : PrimitiveSet[String]
for mapname in tags.keys():
    set = tags.get(mapname)
    assert isinstance(set, PrimitiveSet)
for setname, set in tags:
    assert isinstance(set, PrimitiveSet)

# Test Map with enum key
stamps = ent.Components.add(EnvStampGD).Stamps
stamps.add(RegenerationStateEnum.Dead)
stamps.add(RegenerationStateEnum.Dead)
deadArr = stamps.get(RegenerationStateEnum.Dead)
assert deadArr.node is not None
lushArr = stamps.get(RegenerationStateEnum.Lush)
assert lushArr.node is None
for stampname in stamps.keys():
    stamp = stamps.get(stampname)
    print(stampname)
    assert isinstance(stamp, Array)
for stampname, stamp in stamps:
    print(stampname)
    assert isinstance(stampname, RegenerationStateEnum)
    assert isinstance(stamp, Array)
stamps.remove(RegenerationStateEnum.Dead)
assert stamps.get(RegenerationStateEnum.Dead).node is None

# other
# scp = EntityStateCreatureProfile(None)
# vm = scp.Profile.ValueMap
# vm.get_0()   # Doesn't work yet